### PR TITLE
Allow user to assign value to rubric choices #4746

### LINKED ISSUE #4746: Allow user to assign value to rubric choices
Refer to [issue #4603](https://github.com/TEAMMATES/repo/issues/4603)
Can someone help to figure out a better way for user to assign value to rubric choices?

### PR DESCRIPTION
**Pull request by unyoungwax**
Fixes #4746

Edit: added screenshots
InstructorFeedbackEditPage:
question

InstructorFeedbackResultsPage:
result

### CODE REVIEW

**Comment by kanghj** [comment]
@unyoungwax, post screenshots of your changes.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 1]
```diff
@@ -52,17 +55,26 @@ public boolean extractQuestionDetails(
...
             String choice = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_CHOICE + "-" + i);
-            if(choice != null) {
+            String weight = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHT + "-" + i);
+            if(choice != null && weight != null) {
                 rubricChoices.add(choice);
+
+                try {
+                    rubricWeights.add(Integer.parseInt(weight));
+                } catch (NumberFormatException e) {
+                    rubricWeights.add(0);
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: I don't think catching this is a good idea. I think we should throw an error to the user and show an error message to the user instead if they are supplying non-integer values

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 2]
```diff
@@ -737,5 +800,18 @@ public String getQuestionTypeChoiceOption() {
         return errors;
     }
 
+    private boolean hasAssignedWeights() {
 
+        if (rubricWeights == null || rubricWeights.size() != numOfRubricChoices) {
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: We can remove the test for null, since its initialised to an empty list

File: src/main/webapp/js/instructorFeedbackEdit.js [hunk 3]
```diff
@@ -1550,6 +1571,36 @@ function highlightRubricCol(index, questionNumber, highlight) {
...
+ *            the question number of the feedback question
+ * @returns {Boolean} true if the weights are assigned by the user(the weights
+ *          are not all zeros), otherwise false
+ */
+function hasAssignedWeights(questionNumber) {
+
+    var areWeightsAssigned = false;
+
+    $('#questionTable' + questionNumber).find('input[id^="rubricWeight"]').each(function() {
+        if ($(this).val() != 0) {
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackEdit.js]: <code>!==</code>

Comment (unyoungwax) [inline src/main/webapp/js/instructorFeedbackEdit.js]: `val()` returns a string, should I change the line to `$(this).val() !== '0'`?

Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackEdit.js]: Perhaps try to parseInt ?

**Comment by kanghj** [comment]
On the screenshot for the edit page, did you miss the comment on the issue page to place the "assign weights" button on the top-left cell? Or is there a reason that the current position would work better?

**Comment by unyoungwax** [comment]
I didn't notice it previously, I will move the "assign weights" button to the top-left cell in the next few commits.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 1 above)
Comment (unyoungwax) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Since the method `extractQuestionDetails` does not throw any exception, I will check it in `validateQuestionDetails`.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 4]
```diff
@@ -52,17 +55,26 @@ public boolean extractQuestionDetails(
...
         int numActualChoices = 0;
         int numActualSubQuestions = 0;
         
-        // Get list of choices
+        // Get list of choices and weights
         for(int i = 0 ; i<numOfRubricChoices ; i++) {
             String choice = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_CHOICE + "-" + i);
-            if(choice != null) {
+            String weight = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHT + "-" + i);
+            if(choice != null && weight != null) {
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Just to check, when will this condition ever be false?

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 1 above)
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Yup, then replace this <code>add(0)</code> to an Assumption.fail

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 4 above)
Comment (unyoungwax) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: The condition will be false if the weight or choice is not found in the `requestParameters` (probably due to the user editing the html).

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 5]
```diff
@@ -364,6 +380,11 @@ public String getNewQuestionSpecificEditFormHtml() {
         this.rubricChoices.add("Agree");
         this.rubricChoices.add("Strongly Agree");
         
+        this.rubricWeights.add(0);
+        this.rubricWeights.add(0);
+        this.rubricWeights.add(0);
+        this.rubricWeights.add(0);
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: @damithc, just to check, if the instructor does not assign weights, do we want to give the rubrics choices default weights, or should not display the weights (done in this pr)?

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 6]
```diff
@@ -460,19 +481,27 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
         for (int i = 0; i < numOfRubricChoices; i++) {
             // TODO display numerical value of option 
             String rubricChoiceValue = Sanitizer.sanitizeForHtml(rubricChoices.get(i));
+            String rubricWeight = hasAssignedWeights() ? Integer.toString(rubricWeights.get(i)) : "0";
             String tableHeaderCell = 
                     FeedbackQuestionFormTemplates.populateTemplate(tableHeaderFragmentTemplate,
-                            "${rubricChoiceValue}", rubricChoiceValue);
+                            "${rubricChoiceValue}", rubricChoiceValue + (hasAssignedWeights() ? "(" + rubricWeight + ")" : ""));
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Split this line into 2

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 1 above)
Comment (unyoungwax) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: If `Assumption.fail` is used in the catch-block, the error page will be shown when the weight cannot be parsed. I tried using `add(null)` and then check for null in `validateQuestionDetails`, it works well.

Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Right, I forgot <code>validateQuestionDetails</code> is called only after the question details have been constructed. I'm not very sure what to feel about storing <code>null</code> as a value in the list.

**Comment by unyoungwax** [comment]
I have moved the "assign weights" button to the top-left cell.
question

I have also made some changes so that a warning message is shown when the weights are not integer.
status

**Comment by ashrayjain** [comment]
@unyoungwax Please revert the permissions change to the file ```./gradlew```. This is causing the Travis build for this PR to break. Thanks! :)

**Comment by damithc** [comment]
@unyoungwax No reason to restrict weights to integers right? It should be fine as long as they are numbers?

**Comment by unyoungwax** [comment]
Yes, we can assign floating-point numbers to the weights. How many decimal places should be allowed for the weights?

**Comment by damithc** [comment]
I guess 2 decimal points is reasonable

**Comment by unyoungwax** [comment]
I have made some changes to allow assigning floating-point numbers to weights.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 7]
```diff
@@ -520,10 +557,8 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
         int[][] responseFrequency = calculateResponseFrequency(responses, fqd);
         
         // Initialize percentage frequencies and average value
-        float[][] percentageFrequencyOrAverage = new float[fqd.numOfRubricSubQuestions][];
+        double[][] percentageFrequencyOrAverage = new double[fqd.numOfRubricSubQuestions][fqd.numOfRubricChoices + 1];
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Why <code>+ 1</code>? You can add a comment in the code to explain why

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 8]
```diff
@@ -605,6 +662,11 @@ public String getQuestionResultStatisticsCsv(
                 String percentageFrequency = df.format(rubricStats[i][j] * 100) + "%";
                 csv.append("," + percentageFrequency + " (" + responseFrequency[i][j] + ")");
             }
+
+            if (hasAssignedWeights) {
+                csv.append("," + dfAverage.format(rubricStats[i][numOfRubricChoices]));
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: This <code>rubricStats</code> variable can contain more than one type of value?

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 9]
```diff
@@ -737,5 +804,18 @@ public String getQuestionTypeChoiceOption() {
         return errors;
     }
 
+    private boolean hasAssignedWeights() {
+
+        if (rubricWeights.size() != numOfRubricChoices) {
+            return false;
+        }
+
+        for (Double weight : rubricWeights) {
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Let's just use <code>double</code> here.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 8 above)
Comment (unyoungwax) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: `rubricStats` stores the result of `calculateRubricStats`, an array containing the percentage frequencies and averages.

Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Ok, got it. We can leave it for now, but it looks like it should be done as 2 methods, one for the percentage frequencies and one for average.

File: src/main/webapp/js/instructorFeedbackEdit.js [hunk 10]
```diff
@@ -1550,6 +1572,36 @@ function highlightRubricCol(index, questionNumber, highlight) {
 }
 
 /**
+ * Attaches event handlers to "assign weights" buttons to toggle the visibility
+ * of the input boxes for rubric weights
+ */
+function bindAssignWeightsButtons() {
+    $('body').on('click', 'a[id^="rubricAssignWeightsLink"]', function(e) {
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackEdit.js]: Why is this needed? Doesn't using the selector normally work?

**Comment by kanghj** [comment]
On the edit page, when the "edit" on the question hasn't been clicked, the "assign weights" button is missing. It should instead be there, but disabled, to be consistent with the rest of our UI.

**Comment by kanghj** [comment]
I think we have to add a tooltip to the "assign weights" button and the weights input to describe what the weights are. Starting with default values of 0s for all the options may not be good too.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 11]
```diff
@@ -14,6 +14,7 @@
 public class FeedbackRubricQuestionDetails extends FeedbackQuestionDetails {
     public int numOfRubricChoices;
     public List<String> rubricChoices;
+    public List<Double> rubricWeights;
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Have you tested with legacy data? Will there be problems with old rubrics questions without <code>rubricsWeights</code>?

**Comment by kanghj** [comment]
Btw, even though it's not merged into master yet, do take note of #5024

File: src/main/webapp/js/instructorFeedbackEdit.js (hunk 10 above)
Comment (unyoungwax) [inline src/main/webapp/js/instructorFeedbackEdit.js]: It is easier to maintain the code using this method instead of attaching event handlers directly to the buttons. See #3998

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 11 above)
Comment (unyoungwax) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: I am not sure how to test it correctly. I have no problems editing and showing the results of the rubric question in "Session with different question types"(one of the default sessions in the dev server).

**Comment by unyoungwax** [comment]
If the user clicks the "assign weights" button to hide the weights, should the weights be reset to default or restored to the previous version? Either way we need to change the text of the button to "reset weights" or "restore weights" and add tooltips to explain it.

**Comment by unyoungwax** [comment]
Since the difference between the number of rubric choices and the number of weights is used to determine if there is an empty weight, both values must be equal to edit or add a question successfully. If the user does not assign any weights and the default weights are not all zeros, it is hard to tell whether the user wants to assign weights or not.

**Comment by unyoungwax** [comment]
I made the "assign weights" button to be similar to the "add row" and "add column" button. The buttons are not shown when the "edit" button has not been clicked. Does the "assign weights" button need to have a different behavior?

**Comment by damithc** [comment]
Perhaps a check box 'weighted values' is more appropriate than a button? It should be unticked by default.  
Tick the check box --> shows the text boxes. We can auto-fill some default values (not urgent, can be done later)
Untick the check box --> hides the text boxes. Values are preserved but not used. (if resetting is easier, we can reset for now. Preserving can be done in a future version).

**Comment by unyoungwax** [comment]
question
Is this okay?

**Comment by damithc** [comment]
Placement is fine. I'm not sure 'values' is the right term though. 'Weighted columns'?

**Comment by unyoungwax** [comment]
I will rename "Weighted values" to "Weighted columns". I am not sure if the users know that the term "choices" means rubric choices. I will change the tooltip to "Assign weights to columns to calculate the average".

**Comment by damithc** [comment]
`... calculate average` --> `... calculate statistics`?

**Comment by unyoungwax** [comment]
I have change the input type of "assign weights" from button to checkbox. When the checkbox is ticked, the text boxes are shown and auto-filled with zeros. When the checkbox is unticked, the text boxes are hidden. The weights are not preserved.

File: src/main/java/teammates/common/util/Const.java [hunk 12]
```diff
@@ -290,7 +291,7 @@
         public static final String RUBRIC_ERROR_EMPTY_CHOICE = "Choices for "+Const.FeedbackQuestionTypeNames.RUBRIC + " cannot be empty.";
         public static final String RUBRIC_ERROR_EMPTY_SUB_QUESTION = "Sub-questions for "+Const.FeedbackQuestionTypeNames.RUBRIC + " cannot be empty.";
         public static final String RUBRIC_ERROR_INVALID_CHOICE = "An invalid choice was chosen for the " + Const.FeedbackQuestionTypeNames.RUBRIC + ".";
-        
+        public static final String RUBRIC_ERROR_INVALID_WEIGHT = "The weights for the choices of a " + Const.FeedbackQuestionTypeNames.RUBRIC + " must be valid numbers with precision up to 2 decimal places.";
```
Comment (kanghj) [inline src/main/java/teammates/common/util/Const.java]: I think we can just give a generic failure message here for this.

**Comment by kanghj** [comment]
Perhaps the tooltip should be rephrased to "... to calculate statistics." or "... for calculating statistics.", instead of "... to calculate the statistics."? 

Code looks good to me, but there are some merge conflicts to be resolved now.

**Comment by unyoungwax** [comment]
I have resolved the merge conflict and also change the tooltip of  "assign weights" as suggested.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 13]
```diff
@@ -44,20 +50,23 @@ public boolean extractQuestionDetails(
             FeedbackQuestionType questionType) {
         String numOfRubricChoicesString = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_NUM_COLS);
         String numOfRubricSubQuestionsString = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_NUM_ROWS);
+        String hasAssignedWeightsString = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHTS_ASSIGNED);
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: This line, and a few other lines are too long. wrap?

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 14]
```diff
@@ -66,7 +75,33 @@ public boolean extractQuestionDetails(
...
+            String choice = HttpRequestHelper.getValueFromParamMap(requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_CHOICE + "-" + i);
+
+            if (choice == null) {
+                continue;
+            }
+
+            try {
+                rubricWeights.add(Double.parseDouble(weight));
+            } catch (NumberFormatException e) {
+                // Do not add weight to rubricWeights if the weight cannot be parsed
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Is there some mechanism to prevent/warn users from adding unparsable weights?

File: src/main/webapp/js/instructorFeedbackEdit.js [hunk 15]
```diff
@@ -1561,6 +1581,25 @@ function highlightRubricCol(index, questionNumber, highlight) {
...
+        $(this).closest('form').find('input[id^="rubricWeight"]').toggle();
+    });
+}
+
+/**
+ * @param questionNumber
+ *            the question number of the feedback question
+ * @returns {Boolean} true if the weights are assigned by the user, otherwise false
+ */
+function hasAssignedWeights(questionNumber) {
```
Comment (damithc) [inline src/main/webapp/js/instructorFeedbackEdit.js]: @kanghj I think we still have cases of duplicate question numbers. Can it break things here? Should we use a more trustworthy key to distinguish questions?

**Comment by damithc** [comment]
Don't we need to update instructor help page as well?

**Comment by damithc** [comment]
What happens if a an instructor adds weights to (or modifies weights of) a rubric question after the session has collected some responses?

**Comment by unyoungwax** [comment]
If an instructor assigns or modifies weights, existing responses will not be affected (they will not be deleted).

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 14 above)
Comment (unyoungwax) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: The `type` of the input boxes for the weights is `number`. If the user adds unparsable weights, the browser will prevent form submission and show a warning message. However, it is still possible for the user to assign empty weights, therefore we need to check them in `FeedbackRubricQuestionDetails`.

error
The browser that I am using is Chrome.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 13 above)
Comment (unyoungwax) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: I will wrap the lines like this
```java
String hasAssignedWeightsString = HttpRequestHelper.getValueFromParamMap(
        requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHTS_ASSIGNED);
```

**Comment by unyoungwax** [comment]
I plan to update the instructor help page as follows:

question

statistics

Is it okay?

**Comment by damithc** [comment]
The Weight text box can have a `Weight:` label in front? It doesn't need to be as wide as the text box above it.
How are weights shown in the submission page, and the student results page?

**Comment by unyoungwax** [comment]
The weights are not shown in the submission page. The statistics table is the same for both student results page and instructor feedback results page, i.e. the weights are shown only if the weights are assigned.

**Comment by damithc** [comment]
Can post screenshots of submission and result pages? If the student didn't see weights during submission, won't she be surprised to see them in results?

**Comment by kanghj** [comment]
@damithc, I think we can do the submission page in a different issue?

**Comment by unyoungwax** [comment]
Submission page:
student_submit

Student results page:
student_result

Instructor results page:
instructor_result

I think a student will be confused if he/she sees the weights because he/she does not know what the numbers mean. If the student see the weights in the results page, he/she can know what the numbers mean because there is an "average" column in the table.

**Comment by unyoungwax** [comment]
For the weight labels, is the following okay?
label

**Comment by damithc** [comment]
Um... doesn't look nice. :-p Any other options?

**Comment by unyoungwax** [comment]
I think this is the better way to do it because the content of the table cell will not overflow when there are too many columns. Perhaps we can set the placeholder for the text boxes to "weight" instead of auto-filling the text boxes with zeros?

**Comment by damithc** [comment]
@kanghj I think it is better to do both student and instructor sides together. But enhancements can be done later.
@unyoungwax try to show the word `Weight:` where it is possible. e.g. in student result form

**Comment by unyoungwax** [comment]
I tried to place the text boxes in a new row.

Before and after clicking on the "weights" checkbox:
hidden
row

**Comment by unyoungwax** [comment]
The weight labels in student and instructor results page, are they okay?
statistics

**Comment by damithc** [comment]
Before clicking, show only one row and just the `[ ] Weights`.
After clicking, show both rows and both `Choices --> ` and `[ ] Weights -->`
Yes, the instructor page looks fine. `(...)` part can be made non-bold, but it can be left as a `contributor` issue for future devs.

**Comment by unyoungwax** [comment]
I have added a "weights" row for rubric question in InstructorFeedbackEditPage. "Weight" labels are also added in the student and instructor results page as well as in question statistics csv.

Before and after clicking on the "weights" checkbox:
hidden
row

The results page:
statistics

**Comment by damithc** [comment]
Minor things: there should be a space between the check box and the label. Adding arrows (e.g. Choices --> , Weights --> ) can clarify what the label is referring to.

**Comment by unyoungwax** [comment]
Arrows have been added to the "Choices" and "Weights" table cells.

hidden
row

File: src/main/webapp/js/instructorFeedbackEdit.js (hunk 15 above)
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackEdit.js]: Yes, it will break things here. Most of the existing logic and behaviour in this file relies on question numbers though, we should make a new issue to change the existing logic in this file and in the InstructorFeedbackEditPageData to use the questionId instead.

**Comment by damithc** [comment]
@kanghj can have another look?

File: src/main/webapp/js/instructorFeedbackEdit.js [hunk 16]
```diff
@@ -1446,8 +1457,26 @@ function addRubricCol(questionNumber) {
...
+    rubricWeightFragment = replaceAll(rubricWeightFragment, '${qnIndex}', questionNumber);
+    rubricWeightFragment = replaceAll(rubricWeightFragment, '${col}', newColNumber - 1);
+    rubricWeightFragment = replaceAll(rubricWeightFragment, '${rubricWeight}', 0);
+    rubricWeightFragment = replaceAll(rubricWeightFragment, 
+                                      '${Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHT}',
+                                      'rubricWeight');
+
+    // Insert after last <th>
+    var lastWeightCell = $('#rubricWeights-' + questionNumber + ' th:last');
+    $(rubricWeightFragment).insertAfter(lastWeightCell);
```
Comment (kanghj) [inline src/main/webapp/js/instructorFeedbackEdit.js]: Just to check, although I am not sure what the value of calling <code>disallowNonNumericEntries</code> is, I think to be consistent, we need to call it here too

Comment (damithc) [inline src/main/webapp/js/instructorFeedbackEdit.js]: What's the decision on this one?

File: src/main/webapp/js/instructorFeedbackEdit.js (hunk 15 above)
Comment (damithc) [inline src/main/webapp/js/instructorFeedbackEdit.js]: I don't think we can release this feature if it has the potential to break in production. Hence cannot merge. :-(

**Comment by damithc** [comment]
@kanghj I feel this cannot be merged because it can break during production due to question number issue.  At least there should be an easy way for me to rectify the issue if it happens. What do you think? I occasionally get 'wrong question number' warnings, so that issue is not completely resolved yet and can affect this feature.

**Comment by kanghj** [comment]
We can put this onHold then, until I complete #5208. I don't really think a fix local to this pr is the best way to go about fixing this problem

**Comment by damithc** [comment]
Sure, let's do that. If it's too big to do now, we can do it after exams.

**Comment by damithc** [comment]
@unyoungwax can finish this?

**Comment by unyoungwax** [comment]
I will work on this as soon as #5208 is resolved.

**Comment by kanghj** [comment]
@unyoungwax, we have resolved #5208.

**Comment by unyoungwax** [comment]
I have resolved the conflicts with the master branch.

**Comment by kanghj** [comment]
@unyoungwax, The tests are failing though, do you have time to fix it?

**Comment by unyoungwax** [comment]
I ran all the tests on my computer and they passed. The tests that fail in Travis are not related to this pr (`TimeoutException` in `InstructorCourseStudentDetailsEditPageUiTest` and `NoSuchElementException` in `StudentProfilePageTest`).

**Comment by kanghj** [comment]
Ok, I'll restart the build

**Comment by unyoungwax** [comment]
I have resolved the conflicts with the master branch.

**Comment by wkurniawan07** [comment]
@damithc this one is ready for your review.

File: src/main/java/teammates/common/util/Const.java [hunk 17]
```diff
@@ -316,7 +317,7 @@ public static final String getjQueryUiFilePath(boolean isDevServer) {
         public static final String RUBRIC_ERROR_EMPTY_CHOICE = "Choices for " + Const.FeedbackQuestionTypeNames.RUBRIC + " cannot be empty.";
         public static final String RUBRIC_ERROR_EMPTY_SUB_QUESTION = "Sub-questions for " + Const.FeedbackQuestionTypeNames.RUBRIC + " cannot be empty.";
         public static final String RUBRIC_ERROR_INVALID_CHOICE = "An invalid choice was chosen for the " + Const.FeedbackQuestionTypeNames.RUBRIC + ".";
-        
+        public static final String RUBRIC_ERROR_INVALID_WEIGHT = "The weights for the choices of a " + Const.FeedbackQuestionTypeNames.RUBRIC + " must be valid numbers with precision up to 2 decimal places.";
```
Comment (damithc) [inline src/main/java/teammates/common/util/Const.java]: Line too long

File: src/main/webapp/js/instructorFeedbackEdit.js [hunk 18]
```diff
@@ -1542,6 +1575,59 @@ function highlightRubricCol(index, questionNumber, highlight) {
...
+        var $checkbox = $(this);
+
+        $checkbox.closest('form').find('tr[id^="rubricWeights"]').toggle();
+
+        moveAssignWeightsCheckbox($checkbox);
+    });
+}
+
+/**
+ * Move the "weights" checkbox to the weight row if it is checked, otherwise
```
Comment (damithc) [inline src/main/webapp/js/instructorFeedbackEdit.js]: Moves

**Comment by damithc** [comment]
Minor cosmetic changes requested. In particular, use line wrapping where the github diff view requires horizontal scrolling.

**Comment by unyoungwax** [comment]
I have performed the requested cosmetic changes and also made some changes according to the static analysis.

**Comment by wkurniawan07** [comment]
@kanghj let's review (and hopefully merge) this one quickly before we introduce more changes that might conflict with this PR again.

**Comment by wkurniawan07** [comment]
@damithc ready for your review.

**Comment by wkurniawan07** [comment]
@kanghj I'm merging this now, before more stuffs are added and rendered this PR conflicted again.

@unyoungwax thanks a lot for your perseverance! :+1: :+1:
